bin/peggy.mjs

```diff
@@ -0,0 +1,331 @@
+  }
+
+  const deps = {};
+  for (const dep of options.dependency) {
+    const [name, val] = dep.split(":");
+    deps[name] = val ? val : name;
+  }
+  options.dependencies = deps;
+  delete options.dependency;
+}
+
+if (options.exportVar !== undefined) {
+  if ((options.format !== "globals") && (options.format !== "umd")) {
+    abort(`Can't use the -e/--export-var option with the "${options.format}" module format.`);
+  }
+}
+
+let inputFile = options.input;
+switch (program.args.length) {
+  case 0:
+    inputFile = inputFile ? inputFile : "-";
+    break;
+
+  case 1:
+    if (inputFile) {
+      abort("Do not specify input both on command line and in config file.");
+    }
+    inputFile = program.args[0];
+    break;
+
+  default:
+    abort("Too many arguments.");
+}
+
+let outputFile = options.output;
+if (!outputFile) {
+  outputFile = ((inputFile === "-") || options.test || options.testFile)
+    ? outputFile = "-"
+    : inputFile.substr(0, inputFile.length - path.extname(inputFile).length) + ".js";
+}
+
+if (options.test && options.testFile) {
+  abort("The -t/--test and -T/--test-file options are mutually exclusive.");
+}
+
+options.output = (options.test || options.testFile) ? "parser" : "source";
+let testText = null;
+let testGrammarSource = null;
+if (options.test) {
+  testText = options.test;
+  testGrammarSource = "command line";
+  delete options.test;
+}
+if (options.testFile) {
+  testText = readFile(options.testFile);
+  testGrammarSource = options.testFile;
+  delete options.testFile;
+}
+
+if (verbose) {
+  console.error("OPTIONS:", util.inspect(options, {
+    depth: Infinity,
+    colors: process.stdout.isTTY,
+  }));
+  console.error("INPUT:", inputFile);
+  console.error("OUTPUT:", outputFile);
+  console.error("TEST TEXT:", testText);
+}
+
+// Main
+
+let inputStream, outputStream;
+
+if (inputFile === "-") {
+  process.stdin.resume();
+  inputStream = process.stdin;
+  inputStream.on("error", () => {
+    abort("Can't read from stdin.");
+  });
+  options.grammarSource = "stdin";
+} else {
+  options.grammarSource = inputFile;
+  inputStream = fs.createReadStream(inputFile);
+}
+
+readStream(inputStream, input => {
+  let source;
+
+  try {
+    source = peg.generate(input, options);
+  } catch (e) {
+    abortError("Error parsing grammar:", e);
+  }
+
+  if (testText) {
+    try {
+      source = source.parse(testText, {
+        grammarSource: testGrammarSource,
+      });
+    } catch (e) {
+      abortError("Error parsing test:", e);
+    }
+    source = util.inspect(source, {
+      depth: Infinity,
+      colors: (outputFile === "-") && process.stdout.isTTY,
+    }) + "\n";
+  }
+
+  // Don't create output until processing succeeds
+  if (outputFile === "-") {
+    outputStream = process.stdout;
+  } else {
+    outputStream = fs.createWriteStream(outputFile);
+    outputStream.on("error", () => {
+      abort("Can't write to file \"" + outputFile + "\".");
+    });
+  }
+
+  outputStream.write(source);
```

Comment:
Currently, that will fail/write the wrong results if `--test/--test-file` option is supplied

Reply:
I think it works as expected.  If you specify a test using either mechanism, and you don't specify an output file, you get the output of running the parser on stdout, colorized if stdout is a tty.  If you specify an output file, that's where the output goes.

Reply:
Yes, now I see. I was think that `source` is the parser object at this moment. Would be not what you expect, however. I would expect that `--test` is fully additive, so parser is still generated and writted to the `--output`, and additionally tested against provided input. Instead when the test flag is active, the parser generated only in-memory and test results is written to the `--output`. Not an obvious behaviour if you ask me.

I would suggest the following behaviour:
- if `--output` is explicitly defined, dump parser code to it, even if `--output` is stdout
- when `--test` is defined, the generated parser `eval`uated and tested against input
- test results always written to the stdout
- if `--output` is not explicitly defined and `--test` is defined, do not output the parser source to stdout, but only test results

So, if you just want to test your just-now-written parser, you use only the `--test` option:
```console
$ peggy --test "1" "foo = '1' { return 'success'; }"
success
```
If you want to store parser into file and test in immediately, you specify  the `--output` and `--test` options:
```console
$ peggy --test "1" --output parser.js "foo = '1' { return 'success'; }"
success
...parser is saved to the `parser.js`...
```
if you want to generate parser and do not want to test it and also do not want to write it to the file, you does not specify neither `--output` nor `--test` options:
```console
$ peggy "foo = '1' { return 'success'; }"
...parser code...
```